ping/ping.c

```diff
@@ -1420,8 +1420,13 @@ int ping4_receive_error_msg(struct ping_rts *rts, socket_st *sock)
 
 	res = recvmsg(sock->fd, &msg, MSG_ERRQUEUE | MSG_DONTWAIT);
 	if (res < 0) {
-		if (errno == EAGAIN || errno == EINTR)
+		if (errno == EAGAIN || errno == EINTR) {
 			local_errors++;
+
+			/* SOCK_DGRAM fails on -s > 65527  */
+			if (saved_errno == EMSGSIZE)
+				rts->nerrors++;
+		}
 		goto out;
 	}
 
```

Comment:
Uff this is an ugly hack, so we depend on the fact that the errno is set by the sendto() in ping4_send_probe() but we do not clear the errno before the sendto() so it may have been set by something in the ping initialization code. We really should check if the sendto() returned -1 before we check the errno. Is there a reason why can't we increment the rts->nerrors directly in the send_probe functions?

Reply:
Thanks for catching this. I agree, this is not elegant and errorprone. I guess the handling is then done in `pinger()` function, I would put it there (after `else if (errno == ENOBUFS || errno == ENOMEM)`), it would be on a single place:

https://github.com/iputils/iputils/blob/0f12e6d5bb325df4eb9273b1e294a2cf94a53120/ping/ping_common.c#L361-L380

Here we could even consider early quit (why to keep sending packets which are dropped)?

I suppose the place for validating recvmsg() handled slightly differently for IPv4 and IPv6 is forced by protocol differences.

Reply:
@metan-ucw FYI the error handling moved, let's continue in the new code location.